File: src/main/java/org/example/kalkulationsprogramm/dto/Email/EmailTextTemplateDto.java
package org.example.kalkulationsprogramm.dto.Email;

import com.fasterxml.jackson.annotation.JsonInclude;
import jakarta.validation.constraints.NotBlank;
import java.time.OffsetDateTime;
import lombok.Data;
import org.example.kalkulationsprogramm.domain.EmailTextTemplate;
import org.example.kalkulationsprogramm.domain.EmailTextTemplateKategorie;
import org.example.kalkulationsprogramm.service.EmailTextTemplateKategorien;

@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public class EmailTextTemplateDto {

    private Long id;

    @NotBlank
    private String dokumentTyp;

    /**
     * Fachliche Gruppierung im UI. Optional — wenn das Frontend keinen Wert
     * mitschickt, leitet {@link #applyToEntity(EmailTextTemplate)} die
     * Kategorie aus dem Dokumenttyp ab.
     */
    private EmailTextTemplateKategorie kategorie;

    @NotBlank
    private String name;

    @NotBlank
    private String subjectTemplate;

    @NotBlank
    private String htmlBody;

    private Boolean aktiv;

    private OffsetDateTime createdAt;
    private OffsetDateTime updatedAt;

    public static EmailTextTemplateDto fromEntity(EmailTextTemplate entity) {
        EmailTextTemplateDto dto = new EmailTextTemplateDto();
        dto.setId(entity.getId());
        dto.setDokumentTyp(entity.getDokumentTyp());
        dto.setKategorie(entity.getKategorie() != null
                ? entity.getKategorie()
                : EmailTextTemplateKategorien.kategorieFuer(entity.getDokumentTyp()));
        dto.setName(entity.getName());
        dto.setSubjectTemplate(entity.getSubjectTemplate());
        dto.setHtmlBody(entity.getHtmlBody());
        dto.setAktiv(entity.isAktiv());
        dto.setCreatedAt(entity.getCreatedAt());
        dto.setUpdatedAt(entity.getUpdatedAt());
        return dto;
    }

    public void applyToEntity(EmailTextTemplate entity) {
        if (dokumentTyp != null) {
            entity.setDokumentTyp(dokumentTyp.trim().toUpperCase());
        }
        // Falls das Frontend keine Kategorie mitschickt (z.B. weil die Vorlage
        // mit einer aelteren Frontend-Version angelegt wird), leiten wir sie
        // aus dem Dokumenttyp ab — so bleibt die Gruppierung im UI konsistent
        // ohne dass DB-Zeilen unkategorisiert herumliegen.
        entity.setKategorie(
                kategorie != null
                        ? kategorie
                        : EmailTextTemplateKategorien.kategorieFuer(entity.getDokumentTyp()));
        if (name != null) {
            entity.setName(name.trim());
        }
        if (subjectTemplate != null) {
            entity.setSubjectTemplate(subjectTemplate);
        }
        if (htmlBody != null) {
            entity.setHtmlBody(htmlBody);
        }
        if (aktiv != null) {
            entity.setAktiv(aktiv);
        }
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/dto/Email/EmailTextTemplateDto.kt
package org.example.kalkulationsprogramm.dto.Email

import com.fasterxml.jackson.annotation.JsonInclude
import jakarta.validation.constraints.NotBlank
import org.example.kalkulationsprogramm.domain.EmailTextTemplate
import org.example.kalkulationsprogramm.domain.EmailTextTemplateKategorie
import org.example.kalkulationsprogramm.service.EmailTextTemplateKategorien
import java.lang.reflect.Field
import java.time.OffsetDateTime

@JsonInclude(JsonInclude.Include.NON_NULL)
data class EmailTextTemplateDto(
    var id: Long? = null,
    @field:NotBlank
    var dokumentTyp: String? = null,
    var kategorie: EmailTextTemplateKategorie? = null,
    @field:NotBlank
    var name: String? = null,
    @field:NotBlank
    var subjectTemplate: String? = null,
    @field:NotBlank
    var htmlBody: String? = null,
    var aktiv: Boolean? = null,
    var createdAt: OffsetDateTime? = null,
    var updatedAt: OffsetDateTime? = null,
) {
    fun applyToEntity(entity: EmailTextTemplate) {
        dokumentTyp?.let {
            entity.writeField("dokumentTyp", it.trim().uppercase())
        }
        entity.writeField("kategorie", kategorie ?: EmailTextTemplateKategorien.kategorieFuer(entity.readField("dokumentTyp") as? String))
        name?.let { entity.writeField("name", it.trim()) }
        subjectTemplate?.let { entity.writeField("subjectTemplate", it) }
        htmlBody?.let { entity.writeField("htmlBody", it) }
        aktiv?.let { entity.writeField("aktiv", it) }
    }

    companion object {
        @JvmStatic
        fun fromEntity(entity: EmailTextTemplate): EmailTextTemplateDto {
            val dokumentTyp = entity.readField("dokumentTyp") as? String
            val kategorie = entity.readField("kategorie") as? EmailTextTemplateKategorie
                ?: EmailTextTemplateKategorien.kategorieFuer(dokumentTyp)
            return EmailTextTemplateDto(
                id = entity.readField("id") as? Long,
                dokumentTyp = dokumentTyp,
                kategorie = kategorie,
                name = entity.readField("name") as? String,
                subjectTemplate = entity.readField("subjectTemplate") as? String,
                htmlBody = entity.readField("htmlBody") as? String,
                aktiv = entity.readField("aktiv") as? Boolean,
                createdAt = entity.readField("createdAt") as? OffsetDateTime,
                updatedAt = entity.readField("updatedAt") as? OffsetDateTime,
            )
        }
    }
}

private fun EmailTextTemplate.readField(name: String): Any? {
    val field: Field = EmailTextTemplate::class.java.getDeclaredField(name)
    field.isAccessible = true
    return field.get(this)
}

private fun EmailTextTemplate.writeField(name: String, value: Any?) {
    val field: Field = EmailTextTemplate::class.java.getDeclaredField(name)
    field.isAccessible = true
    field.set(this, value)
}
